File: src/utils/stableId.ts
/** Deterministic id from path segments — stable across reloads and sync. */
export function stableIdFromParts(...parts: string[]): string {
  const input = parts.filter(Boolean).join('\0');
  let hash = 2166136261;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  const hex = (hash >>> 0).toString(16).padStart(8, '0');
  return `tn_${hex}`;
}

export function stableIdFromPath(path: string): string {
  return stableIdFromParts(path);
}

export function stableNoteBlockId(notebookPath: string, index: number, createdAt: string): string {
  return stableIdFromParts(notebookPath, String(index), createdAt);
}

File: mobile/lib/core/stable_id.dart
String stableIdFromParts(List<String> parts) {
  final input = parts.where((part) => part.isNotEmpty).join('\u0000');
  var hash = 2166136261;
  for (var i = 0; i < input.length; i++) {
    hash ^= input.codeUnitAt(i);
    hash = (hash * 16777619) & 0xFFFFFFFF;
  }
  final hex = hash.toRadixString(16).padLeft(8, '0');
  return 'tn_$hex';
}

String stableIdFromPath(String path) => stableIdFromParts([path]);

String stableNoteBlockId(String notebookPath, int index, String createdAt) {
  return stableIdFromParts([notebookPath, index.toString(), createdAt]);
}
